src/google/protobuf/repeated_field.h

```diff
@@ -1255,10 +1255,12 @@ void RepeatedField<Element>::Reserve(int new_size) {
   }
   // Likewise, we need to invoke destructors on the old array. If Element has no
   // destructor, this loop will disappear.
-  e = &old_rep->elements[0];
-  limit = &old_rep->elements[current_size_];
-  for (; e < limit; e++) {
-    e->Element::~Element();
+  if (current_size_ > 0) {
```

Comment:
Careful! If `Element::~Element()` is not visible to the compiler, it implies a memory barrier, so the `old_rep->elements` fetch is not loop-invariant, so it can't be hoisted out of the loop. That's the reason for the `e` and `limit` pointers pre-computed above the loop. This code is *very* sensitive for both code size and performance, so please verify equivalent machine code for types with nontrivial destructors.